### ISSUE TITLE
bug: missing commits in file history panel when following renames

### ISSUE THREAD

**Issue by farbodsz** [sindrets/diffview.nvim#66]
When running `:DiffviewFileHistory` on a file that has been renamed/moved, I'm not getting commits before the rename commit in the file history panel (only a few commits):

```
M | ... ... | some change
A | ... ... | move file to ...
```

However if I run `!git log --follow %`, then I get all the commits after the rename commit as expected (around 60). I get the same output of around 60 commits after running `:0Gclog` (from `vim-fugitive`).

Here's my configuration:

```lua
 local cb = require("diffview.config").diffview_callback

require("diffview").setup({
  diff_binaries = false,
  use_icons = false,

  file_panel = {
    position = "left",
    width = 40,
  },

  file_history_panel = {
    position = "bottom",
    height = 16,
    follow = true,
  },

  key_bindings = {
    disable_defaults = false,
    file_panel = {
      ["k"] = cb("prev_entry"),
      ["j"] = cb("next_entry"),
      ["[f"] = cb("select_prev_entry"),
      ["]f"] = cb("select_next_entry"),
      ["<leader>t"] = cb("toggle_files"),
    },
    file_history_panel = {
      ["k"] = cb("prev_entry"),
      ["j"] = cb("next_entry"),
      ["[f"] = cb("select_prev_entry"),
      ["]f"] = cb("select_next_entry"),
      ["<leader>t"] = cb("toggle_files"),
      ["g?"] = cb("options"),
      ["zC"] = cb("close_all_folds"),
      ["zM"] = cb("close_all_folds"),
      ["zO"] = cb("open_all_folds"),
      ["zR"] = cb("open_all_folds"),
      ["<leader>gD"] = cb("open_in_diffview"),
    },
    option_panel = {
      ["<return>"] = cb("select"),
      ["q"] = cb("close"),
    },
    view = {
      ["[f"] = cb("select_prev_entry"),
      ["]f"] = cb("select_next_entry"),
      ["<leader>t"] = cb("toggle_files"),
    },
  },
})
```

I'll take a closer look at this when I get home, though I'm wondering if you know what might be causing this?

**Comment by sindrets** [comment]
`--follow` is not enabled by default. But it can be enabled from the file history option panel. Looking at your config I can see that you have mapped this to `g?` from the file history panel.

Also, looking at this:

```
file_history_panel = {
    position = "bottom",
    height = 16,
    follow = true,
  },
  ```
  
  Seems like you have some of the early WIP configuration here :) All the log options were moved to a new object:
  
  ```
  file_history_panel = {
    position = "bottom",
    width = 35,
    height = 16,
    log_options = {
      max_count = 256,      -- Limit the number of commits
      follow = false,       -- Follow renames (only for single file)
      all = false,          -- Include all refs under 'refs/' including HEAD
      merges = false,       -- List only merge commits
      no_merges = false,    -- List no merge commits
      reverse = false,      -- List commits in reverse order
    },
  ```

**Comment by sindrets** [comment]
@farbodsz Can I consider this resolved, or do you have any other questions?

**Comment by farbodsz** [comment]
Hey @sindrets  - sorry for replying so late! I've been super busy this week so haven't had a chance to provide further details.

> Seems like you have some of the early WIP configuration here :) All the log options were moved to a new object:

Haha yes, good catch. It must've been from when I was using the WIP feat branch earlier.

But even with this configuration I'm still having some issues with diffview not showing the full file history in the cases I mentioned, so if you could leave this open for now that would be great.

I can properly look into it and give you the details tomorrow/Thursday, if that's ok?

**Comment by sindrets** [comment]
Of course, no worries.

I think I know what's happening here though. `follow` in diffview also implies `--first-parent`. There are several reasons for this. Try to compare the output of `git log --follow --first-parent -- path/to/file`, to the entries in file history view with `follow` enabled. They should be the same.

As for why I use `--first-parent`: 
`--follow` forces git log to traverse all commits it can find by following all commit parents. This means that for every merge commit, the history will branch. This is fine for something like fugitive's `:Gclog`, but it makes a whole lot less sense for diffview's file history. Because in the file history view, the goal is to show the incremental change of a file. When listing all these commits from all connected branches in chronological order, there is no longer any guarantee that any one commit is related to the previous or the next.

In short: presenting all these commits in a list would suggest that they are all part of a linear history. But that is not necessarily true, and is simply very misleading. I looked to other git file history tools while I was developing this, and they all seem to be doing the same thing with regards to following renames (i.e. the VSCode extension [GitLens](https://github.com/eamodio/vscode-gitlens)).

**Comment by farbodsz** [comment]
Yeah you're right - I think I initially tried playing around with the `--first-parent` option in the diffview source when I encountered this problem as well.

The behaviour I was expecting was the one without `--first-parent` but I can see now how including it makes much more sense for diffview's purpose.
